index_impl.go

```diff
@@ -460,16 +498,31 @@ func (i *indexImpl) SearchInContext(ctx context.Context, req *SearchRequest) (sr
 		coll = collector.NewTopNCollector(req.Size, req.From, req.Sort)
 	}
 
-	// open a reader for this search
-	indexReader, err := i.i.Reader()
-	if err != nil {
-		return nil, fmt.Errorf("error opening index reader %v", err)
+	var knnHits []*search.DocumentMatch
+	var ok bool
+	var skipKnnCollector bool
+	if req.Metadata != nil {
+		for k, v := range req.Metadata {
+			switch k {
+			case search.KnnMetadataKey:
+				if v != nil {
+					knnHits, ok = v.([]*search.DocumentMatch)
+					if !ok {
+						return nil, fmt.Errorf("knn metadata must be of type []*search.DocumentMatch")
+					}
+				}
+				skipKnnCollector = true
+			}
+		}
 	}
-	defer func() {
-		if cerr := indexReader.Close(); err == nil && cerr != nil {
-			err = cerr
+	if !skipKnnCollector && requestHasKNN(req) {
```

Comment:
a question over here, is this block hit when you run the KNN query on a single bleve index?

Reply:
this block gets hit only when 
- It is not presearch which means 
     - It is in phase 2 of a 2 phase search or 
     - It is a simple query with no 2 phase requirement
- If it is in phase 2 of a 2 phase search then Request will have the PresearchData/Metadata so you just use that and dont search for knn